hist/histv7/inc/ROOT/RHistData.hxx

```diff
@@ -257,44 +298,76 @@ public:
+   Weight_t GetBinArray(int binidx) const
+   {
+      if (binidx < 0){
+         return fOverflowSumWeightsSquared[-binidx - 1];
+      } else {
+         return fSumWeightsSquared[binidx - 1];
+      }
+   }
+
+   /// Get a reference to the bin corresponding to `binidx` of the correct bin
+   /// content array (non-const)
+   /// i.e. depending if `binidx` is a regular bin or an under- / overflow bin.
+   Weight_t& GetBinArray(int binidx)
+   {
+      if (binidx < 0){
+         return fOverflowSumWeightsSquared[-binidx - 1];
+      } else {
+         return fSumWeightsSquared[binidx - 1];
+      }
+   }
+
+   /// Add weight to the bin at `binidx`; the coordinate was `x`.
    void Fill(const CoordArray_t & /*x*/, int binidx, Weight_t weight = 1.)
    {
-      fSumWeightsSquared[binidx] += weight * weight;
+      GetBinArray(binidx) += weight * weight;
    }
 
    /// Calculate a bin's (Poisson) uncertainty of the bin content as the
    /// square-root of the bin's sum of squared weights.
-   double GetBinUncertaintyImpl(int binidx) const { return std::sqrt(fSumWeightsSquared[binidx]); }
+   double GetBinUncertaintyImpl(int binidx) const { return std::sqrt(GetBinArray(binidx)); }
 
    /// Get a bin's sum of squared weights.
-   Weight_t GetSumOfSquaredWeights(int binidx) const { return fSumWeightsSquared[binidx]; }
-
+   Weight_t GetSumOfSquaredWeights(int binidx) const { return GetBinArray(binidx); }
    /// Get a bin's sum of squared weights.
-   Weight_t &GetSumOfSquaredWeights(int binidx) { return fSumWeightsSquared[binidx]; }
+   Weight_t &GetSumOfSquaredWeights(int binidx) { return GetBinArray(binidx); }
 
    /// Get the structure holding the sum of squares of weights.
    const std::vector<double> &GetSumOfSquaredWeights() const { return fSumWeightsSquared; }
    /// Get the structure holding the sum of squares of weights (non-const).
    std::vector<double> &GetSumOfSquaredWeights() { return fSumWeightsSquared; }
 
-   /// Merge with other RHistStatUncertainty data, assuming same bin configuration
+   /// Get the structure holding the under-/overflow sum of squares of weights.
+   const std::vector<double> &GetOverflowSumOfSquaredWeights() const { return fOverflowSumWeightsSquared; }
+   /// Get the structure holding the under-/overflow sum of squares of weights (non-const).
+   std::vector<double> &GetOverflowSumOfSquaredWeights() { return fOverflowSumWeightsSquared; }
+
+   /// Merge with other `RHistStatUncertainty` data, assuming same bin configuration.
    void Add(const RHistStatUncertainty& other) {
-      assert(fSumWeightsSquared.size() == other.fSumWeightsSquared.size());
+      assert(fSumWeightsSquared.size() + fOverflowSumWeightsSquared.size() == other.fSumWeightsSquared.size() + other.fOverflowSumWeightsSquared.size());
       for (size_t b = 0; b < fSumWeightsSquared.size(); ++b)
          fSumWeightsSquared[b] += other.fSumWeightsSquared[b];
+      for (size_t b = 0; b < fOverflowSumWeightsSquared.size(); ++b)
```

Comment:
Slight improvement (a pattern):

```suggestion
      for (size_t b = 0, e = fOverflowSumWeightsSquared.size(); b < e; ++b)
```

Else we have an extra indirections, see https://www.godbolt.org/z/LYaswm `.L3` block vs `.L12` block in the assembly.

Reply:
I'm not familiar with this problem, could you explain a bit more what this is about, please?

Reply:
I would guess that the compiler fails to prove that the subsequent operations will not modify the vector's size, and therefore has to constantly re-check the vector size on every loop iteration unless said vector size is manually cached.

One reason why this may happen (not sure if that's the actual reason) is that in C/++, a pointer can point to pretty much everything, and therefore the compiler cannot prove that `vector[index]` doesn't actually point to the vector struct's inner size field.